crates/functional-yew/src/lib.rs

```diff
@@ -0,0 +1,507 @@
+where
+    InitialProvider: FnOnce() -> T,
+{
+    struct UseRefState<T> {
+        o: Rc<RefCell<T>>,
+    }
+
+    let initial_value = Box::new(initial_value);
+
+    use_hook(
+        |state: &mut UseRefState<T>, pretrigger_change_acceptor| {
+            let _ignored = || pretrigger_change_acceptor(|_| false); // we need it to be a specific closure type, even if we never use it
+            return state.o.clone();
+        },
+        move || UseRefState {
+            o: Rc::new(RefCell::new(initial_value())),
+        },
+    )
+}
+
+pub fn use_reducer1<Action: 'static, Reducer, State: 'static>(
+    reducer: Reducer,
+    initial_state: State,
+) -> (Rc<State>, Box<impl Fn(Action)>)
+where
+    Reducer: Fn(Rc<State>, Action) -> State + 'static,
+{
+    return use_reducer2(reducer, initial_state, |a| a);
+}
+
+pub fn use_reducer2<Action: 'static, Reducer, State: 'static, InitialState, InitFn>(
+    reducer: Reducer,
+    initial_state: InitialState,
+    init: InitFn,
+) -> (Rc<State>, Box<impl Fn(Action)>)
+where
+    Reducer: Fn(Rc<State>, Action) -> State + 'static,
+    InitFn: Fn(InitialState) -> State,
+{
+    struct UseReducerState<State> {
+        current_state: Rc<State>,
+    }
+    let init = Box::new(init);
+    let reducer = Rc::new(reducer);
+    let ret = use_hook(
+        |internal_hook_change: &mut UseReducerState<State>, pretrigger_change_runner| {
+            return (
+                internal_hook_change.current_state.clone(),
+                Box::new(move |action: Action| {
+                    let reducer = reducer.clone();
+                    pretrigger_change_runner(
+                        move |internal_hook_change: &mut UseReducerState<State>| {
+                            internal_hook_change.current_state = Rc::new((reducer)(
+                                internal_hook_change.current_state.clone(),
+                                action,
+                            ));
+                            true
+                        },
+                    );
+                }),
+            );
+        },
+        move || UseReducerState {
+            current_state: Rc::new(init(initial_state)),
+        },
+    );
+    return ret;
+}
+
+pub fn use_state<T, F>(initial_state_fn: F) -> (Rc<T>, Box<impl Fn(T)>)
+where
+    F: FnOnce() -> T,
+    T: 'static,
+{
+    struct UseStateState<T2> {
+        current: Rc<T2>,
+    }
+    return use_hook(
+        |prev: &mut UseStateState<T>, hook_update| {
+            let current = prev.current.clone();
+            return (
+                current,
+                Box::new(move |o: T| {
+                    hook_update(|state: &mut UseStateState<T>| {
+                        state.current = Rc::new(o);
+                        true
+                    });
+                }),
+            );
+        },
+        move || UseStateState {
+            current: Rc::new(initial_state_fn()),
+        },
+    );
+}
+
+pub fn use_effect<F, Destructor>(callback: F)
```

Comment:
We may want to turn this into a macro, but let's save it for later

Reply:
Agreed :)